tiledb/sm/query/query_ast.h

```diff
@@ -0,0 +1,214 @@
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Defines the Query AST node classes.
+ */
+
+#ifndef TILEDB_QUERY_AST_H
+#define TILEDB_QUERY_AST_H
+
+#include <iostream>
+#include <sstream>
+
+#include "tiledb/common/common.h"
+#include "tiledb/common/status.h"
+#include "tiledb/sm/array_schema/array_schema.h"
+#include "tiledb/sm/enums/query_condition_combination_op.h"
+#include "tiledb/sm/enums/query_condition_op.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+enum ASTNodeTag : char { NIL, VAL, EXPR };
+
+/** Represents a simple terminal/predicate **/
+class ASTNode {
+ public:
+  virtual ASTNodeTag get_tag() = 0;
+  virtual std::string to_str() = 0;
+};
+class ASTNodeVal : public ASTNode {
+  /** Value constructor. */
+ public:
+  ASTNodeVal(
+      std::string field_name,
+      const void* const condition_value,
+      const uint64_t condition_value_size,
+      const QueryConditionOp op)
+      : field_name_(field_name)
+      , op_(op) {
+    condition_value_data_.resize(condition_value_size);
+
+    if (condition_value != nullptr) {
+      // Using an empty string literal for empty string as vector::resize(0)
+      // doesn't guarantee an allocation.
+      memcpy(
+          condition_value_data_.data(), condition_value, condition_value_size);
+    }
+  }
+
+  virtual ~ASTNodeVal() {
+  }
+
+  /** Copy constructor. */
+  ASTNodeVal(const ASTNodeVal& rhs)
+      : field_name_(rhs.field_name_)
+      , condition_value_data_(rhs.condition_value_data_)
+      , op_(rhs.op_) {
+  }
+
+  /** Move constructor. */
+  ASTNodeVal(ASTNodeVal&& rhs)
+      : field_name_(std::move(rhs.field_name_))
+      , condition_value_data_(std::move(rhs.condition_value_data_))
+      , op_(rhs.op_) {
+  }
+
+  /** Assignment operator. */
+  ASTNodeVal& operator=(const ASTNodeVal& rhs) {
+    if (this != &rhs) {
+      field_name_ = rhs.field_name_;
+      condition_value_data_ = rhs.condition_value_data_;
+      op_ = rhs.op_;
+    }
+
+    return *this;
+  }
+
+  /** Move-assignment operator. */
+  ASTNodeVal& operator=(ASTNodeVal&& rhs) {
+    field_name_ = std::move(rhs.field_name_);
+    condition_value_data_ = std::move(rhs.condition_value_data_);
+    op_ = rhs.op_;
+
+    return *this;
+  }
+
+  ASTNodeTag get_tag() {
+    return tag_;
+  }
+
+  std::string to_str() {
+    std::string result_str;
+    result_str = field_name_ + " " + query_condition_op_str(op_) + " ";
+    result_str += condition_value_data_.to_hex_str();
+    return result_str;
+  }
+
+  /** The attribute name. */
+  std::string field_name_;
+
+  /** The value data. */
+  ByteVecValue condition_value_data_;
+
+  /** The comparison operator. */
+  QueryConditionOp op_;
+
+ private:
+  static const ASTNodeTag tag_{ASTNodeTag::VAL};
+};
+
+class ASTNodeExpr : public ASTNode {
+ public:
+  ASTNodeExpr(
+      std::vector<shared_ptr<ASTNode>> nodes, QueryConditionCombinationOp c_op)
```

Comment:
The reason I am using a shared pointer is because this gives an easier way of casting between classes, and I didn't want to use regular pointers because allocating and deallocating is a pain. I don't think it is possible to do the inheritance casting with just a vector of ASTNodes.

Reply:
If we want to implement copy construction we should not use unique_ptr, is there any reason to not implement copy construction?

Reply:
Is there a reason *to* implement a copy constructor?  If so then a follow-up question is whether the copy should be shallow or deep.  In the case of shallow copies, a shared_ptr would be necessary, in the case of deep copies, a unique_ptr would work.  Since the ASTNodeExpr has a vector of nodes -- and vector is deep copy (the ptr is being used for polymorphism, not for shallow-ness) -- deep copy is probably what we want?  If we want copy at all.

Reply:
Copying may be necessary for AST optimizations, eg we might need to copy a node and all of its children somewhere else when doing an AST rewrite [^1]. Deep copying would make the most sense in that case, I think.

[^1] There are use-cases for keeping the original AST form while doing the 'actual work' with an optimized copy.